### ISSUE TITLE
[FR] [h]ibernate option

### ISSUE THREAD

**Issue by Shooter3k** [hashcat/hashcat#2890]
It would be great to add a hibernate option. 

This option would allow a person to hit H and 'hibernate' hashcat. The hibernate option would be similar to pause except it would remove the data from gpu memory and put hashcat into a CPU idle status and possibly limit the ram it's using.  This option would be very useful when running multiple hashcat sessions concurrently and you need the GPU memory to be used on certain sessions.  The assumption is this would be different than "pause" (immediate resume) as some time may be needed to reload data back into things once the session was resumed. 

Possible implementation might look like: 
[s]tatus [p]ause [b]ypass [c]heckpoint [f]inish **[h]hibernate** [q]uit =>

Currently hashcat does remove the data from GPU memory (sometimes, depending on your driver?) if you pause the current attack, open a new hashcat session and do a small benchmark.

**Comment by roycewilliams** [comment]
This could work similarly to what happens when quitting and restoring later, but without all of the overhead of doing that manually. I think it could be pretty useful.

**Comment by Chick3nman** [comment]
While I can see how this could be a little useful, I'm not sure this would be worthwhile to add as its own function. This would leave hashcat in a bit of an odd place, essentially the same as quitting with `q` from what I understand but still running, which could be a bit more complicated than it may seem on the surface. We would need to run all of the initialization stuff over again, as pointed out in the initial post, and we would still be restarting from where the last checkpoint was taken since it doesn't sound like this is emulating the checkpoint quit functionality and waiting for the next checkpoint before suspending operation. Do we leave any of the timers running in the original process? Do we clean everything up like with a `quit` or do we leave some files in place? Would this keep the mutex/lock on the default hashcat session? If so, the user is still responsible for setting sessions themselves here to run the additional sessions, meaning the overhead they save with this is simply in running `hashcat --restore`.

**Comment by Shooter3k** [comment]
When working on hash lists it seems very very common for people to have a big list running and then want to squeeze in some 'let me check this a second' attacks. If a person stops the large attacks a lot of time is lost waiting for things to gracefully stop and start back up.

The primary objective here is to release the data from gpu memory so a quick resume is possible without having to restart everything. It appears it's already happening on it's own when you do a hashcat -b when another session is paused (then when you resume your machine will usually lock up for a couple of seconds while the data is copied back into gpu memory). My idea was to implement this but in a less hacky fashion. If other optimizations are possible at the same time then great, if not, that's ok too. 

tldr; running a hashcat -b to clear out gpu memory is a total hack and it seems like we can do it better

**Comment by jsteube** [comment]
From a technical perspective this is not fully thought out. The data has to go somewhere to load it on resume, even if that is just temporarily, but it this place can't be the host memory, because there's many scenarios (especially in professional use) where we have more GPU memory than host memory. IOW, is has to go to disc. But if it goes to disc, there's no advantage over using quit and restore as @Chick3nman described.

**Comment by roycewilliams** [comment]
Most beginning users would not need this feature. But most power users will eventually reach a point where it will make sense to write a script to wrap hashcat to recreate this feature outside hashcat - because it will save time and improve consistency in "accuracy" (of being able to successfully and quickly "resume" an attack, without having to track session names).

I personally would love this being available in hashcat, because it would allow me to remove that complexity from my own personal scripts and workflow. But I can also see where the project might not want to take on that complexity, and instead leave that complexity to the few users that need it. At that point, it becomes a competitive advantage that is only available to the users capable of scripting it.

As always, when there is a way to free up the user to be able to focus more on the fun parts of cracking, I will always advocate that hashcat take on that complexity on behalf of the user. 😁 But I totally respect the opposite decision as well. And a script wrapper to do this is not very hard.

But no script wrapper would be able to do this as quickly, because restoring requires reparsing hashes and all of the other overhead that takes time when starting hashcat. That's where doing it in hashcat would provide value that no script can. If hashcat does it within the existing process, hashcat can trust the restore data more and skip a number of time-consuming validation steps. This would provide a significant amount of flexibility and speed when managing multiple non-trivial jobs in parallel, that I would personally use the heck out of!

I would say to make it safe to use, to restrict the user to not be able to use hibernate when the shared common potfile is used, and the inputs (hashes, word list, etc) must not have changed or must be ignored and only internal restoration data be used. (A previous version of this post said that it should also not be allowed if the potfile has changed, but I have changed my mind on this. It is the inputs that must not change, not the outputs, though the potfile is a special case where it is also an input)
